File: netpulse-api/internal/geo/privacy.go
package geo

import (
	"math"
	"strconv"
)

// SnapDegree rounds a coordinate to 1 degree (~100 km). Precise points are never returned.
func SnapDegree(value float64) float64 {
	return math.Round(value)
}

func CoarsePoint(lon, lat float64) (float64, float64, bool) {
	if lon < -180 || lon > 180 || lat < -90 || lat > 90 {
		return 0, 0, false
	}
	if math.IsNaN(lon) || math.IsNaN(lat) || math.IsInf(lon, 0) || math.IsInf(lat, 0) {
		return 0, 0, false
	}
	return SnapDegree(lon), SnapDegree(lat), true
}

func InViewport(lon, lat float64, q Query) bool {
	if !q.HasViewport() {
		return true
	}
	west, south, east, north := *q.West, *q.South, *q.East, *q.North
	if lat < south || lat > north {
		return false
	}
	if west <= east {
		return lon >= west && lon <= east
	}
	return lon >= west || lon <= east
}

func ClusterKey(lon, lat float64) string {
	return strconv.Itoa(int(SnapDegree(lon))) + ":" + strconv.Itoa(int(SnapDegree(lat)))
}

File: netpulse-web/src/features/map/privacy.ts
import type { MapViewport } from "@/domain/map";

export function snapDegree(value: number): number {
  return Math.round(value);
}

export function isCoarseCoordinate(
  lon: number | null | undefined,
  lat: number | null | undefined
): boolean {
  if (lon == null || lat == null) {
    return false;
  }
  if (!Number.isFinite(lon) || !Number.isFinite(lat)) {
    return false;
  }
  if (lon < -180 || lon > 180 || lat < -90 || lat > 90) {
    return false;
  }
  return Number.isInteger(lon) && Number.isInteger(lat);
}

export function roundViewport(viewport: MapViewport): MapViewport {
  return {
    west: snapDegree(viewport.west),
    south: snapDegree(viewport.south),
    east: snapDegree(viewport.east),
    north: snapDegree(viewport.north),
  };
}

export function stripPreciseCoordinates<T extends { lon: number | null; lat: number | null }>(
  cell: T
): T {
  if (!isCoarseCoordinate(cell.lon, cell.lat)) {
    return { ...cell, lon: null, lat: null };
  }
  return cell;
}
